### PR TITLE
Initial commit of flight recorder trace

### PR DESCRIPTION
**Pull request by c-p-i-o**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130764

Summary:
`fr_trace.py` is used to analyze flight recorder dump files.
This script was taken from @wconstab and @zdevito. 
Only minor changes made were to make the linter happy and add a few odd new fields that I added in version `2.2` of the collector portions. 

Test Plan:
Tested manually on some flight recorder data and it seems to run.

TODO:
Address 15 odd `#type: ignore` that I put in there to make the linter happy for now.

### CODE REVIEW

File: tools/flight_recorder/fr_trace.py [hunk 1]
```diff
@@ -0,0 +1,831 @@
...
+        sys.exit(0)
+
+    tracebacks, collectives, nccl_calls = build_collectives(
+        entries, _groups, _memberships
+    )
+    print("built collectives, nccl_calls")
+    # print("Groups\n", tabulate(groups, headers=Group._fields))
+    # print("Memberships\n", tabulate(memberships, headers=Membership._fields))
+    # print("Collectives\n", tabulate(collectives, headers=Collective._fields))
+    # print("NCCLCalls\n", tabulate(nccl_calls, headers=NCCLCall._fields))
```
Comment (fduwjj) [inline tools/flight_recorder/fr_trace.py:753]: Do we want to keep this?

File: tools/flight_recorder/fr_trace.py [hunk 2]
```diff
@@ -0,0 +1,831 @@
...
+    basename = os.path.basename(filename)
+    assert (
+        basename.find(prefix) == 0
+    ), f"args.prefix ({prefix}) must match the beginning of each filename ({basename})"
+    rank = int(basename[len(prefix) :])
+    #     host_name, _, rank = os.path.basename(filename).split("_", 2)
+    # except ValueError:
+    #     _, rank = os.path.basename(filename).split("_", 1)
+    #     host_name = "host0"
+    # rank = int(rank)
```
Comment (fduwjj) [inline tools/flight_recorder/fr_trace.py:774]: ditto

**Review by fduwjj — APPROVED** [review]
Stamp and please make sure the linter is happy, thanks!

File: tools/flight_recorder/fr_trace.py (hunk 1 above)
Comment (c-p-i-o) [reply tools/flight_recorder/fr_trace.py:753]: This is useful. I'll add a `--debug` switch and put all these behind the `debug` global switch.

File: tools/flight_recorder/fr_trace.py (hunk 2 above)
Comment (c-p-i-o) [reply tools/flight_recorder/fr_trace.py:774]: will remove!

**Pull request closed without merging**